### PR TITLE
Homework make_unique

### PR DESCRIPTION
**Pull request by jakjan95**
Authors: @jakjan95

### CODE REVIEW

File: module1/homework/make_unique/include/make_unique.hpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <memory>
+#include <type_traits>
+
+namespace cs {
+template <typename T,
+          typename... Args,
+          typename = std::enable_if_t<!std::is_array_v<T>>>
```
Comment (majkel84) [inline module1/homework/make_unique/include/make_unique.hpp:9]: great idea to pass it as template argument

File: module1/homework/make_unique/tests/make_unique.ut.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+};
+
+TEST(MakeUniqueTest, ShouldCreateUniquePointerFromInt) {
+    auto ptr = cs::make_unique<int>(initialValue);
+    ASSERT_EQ(*ptr, initialValue);
+}
+
+TEST(MakeUniqueTest, ShouldCreateUniquePointerFromCustomClassWithDefaultValue) {
+    auto testClass = cs::make_unique<TestClass>();
+    ASSERT_EQ(testClass->getValue(), defaultValuesOfTestClass);
```
Comment (majkel84) [inline module1/homework/make_unique/tests/make_unique.ut.cpp]: why get value in here, *testCalss should do the same

File: module1/homework/make_unique/include/make_unique.hpp (hunk 1 above)
Comment (MaciejSalamonski) [reply module1/homework/make_unique/include/make_unique.hpp:9]: +1

**Review by MaciejSalamonski — APPROVED** [review]
Good job @jakjan95 👍

File: module1/homework/make_unique/include/make_unique.hpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+auto make_unique(Args&&... args) {
+    return std::unique_ptr<T>{new T(std::forward<Args>(args)...)};
+}
+
+template <typename T>
+auto make_unique(size_t size) {
+    return std::unique_ptr<T>{new std::remove_extent_t<T>[size] {}};
+}
```
Comment (ziobron) [inline module1/homework/make_unique/include/make_unique.hpp:17]: I'm curious why it doesn't leak. This should be a proper version for arrays.

```suggestion
template <typename T>
auto make_unique(size_t size) {
    return std::unique_ptr<T[]>{new std::remove_extent_t<T>[size] {}};
}
```

File: module1/homework/make_unique/tests/make_unique.ut.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+public:
+    TestClass(int value = defaultValuesOfTestClass)
+        : value_{value} {}
+
+    int getValue() const {
+        return value_;
+    }
+
+private:
+    int value_;
```
Comment (ziobron) [inline module1/homework/make_unique/tests/make_unique.ut.cpp]: why not make it public? It's for test purpose only.

**Review by ziobron — CHANGES_REQUESTED** [review]
you have tests only for 1 and 0 arguments. Please add a class that can take more arguments.
There are no tests assuring that proper constructors with r-values and l-values are called.

**Review by WBigaj — COMMENTED** [review]
I have checked code and have nothing to add that was not mentioned by previous reviewers.

File: module1/homework/make_unique/include/make_unique.hpp (hunk 3 above)
Comment (jakjan95) [reply module1/homework/make_unique/include/make_unique.hpp:17]: T is array type as we passing it by template :)

**Comment by jakjan95** [comment]
Added tests for two arguments also added test cases for left and right values constructors :)

File: module1/homework/make_unique/include/make_unique.hpp (hunk 3 above)
Comment (ziobron) [reply module1/homework/make_unique/include/make_unique.hpp:17]: you are right, my mistake.

File: module1/homework/make_unique/tests/make_unique.ut.cpp [hunk 5]
```diff
@@ -45,3 +72,14 @@ TEST(MakeUniqueTest, ShouldCreateUniquePointerForArrayType) {
     ASSERT_EQ(arrayPtr[0], initialValue);
     ASSERT_EQ(arrayPtr[1], secondValue);
 }
+
+TEST(MakeUniqueTest, ShouldCreatePointerUsingLeftValue) {
+    auto sideTestClassPtr = cs::make_unique<LRValueTestClass>();
+    ASSERT_EQ(sideTestClassPtr->sideTestValue_, leftValue);
+}
```
Comment (ziobron) [inline module1/homework/make_unique/tests/make_unique.ut.cpp]:
```suggestion
TEST(MakeUniqueTest, ShouldCreatePointerUsingCopyConstructor) {
    LRValueTestClass copy;
    auto sideTestClassPtr = cs::make_unique<LRValueTestClass>(copy);
    ASSERT_EQ(sideTestClassPtr->sideTestValue_, leftValue);
}
```

File: module1/homework/make_unique/tests/make_unique.ut.cpp [hunk 6]
```diff
@@ -45,3 +72,14 @@ TEST(MakeUniqueTest, ShouldCreateUniquePointerForArrayType) {
...
+
+TEST(MakeUniqueTest, ShouldCreatePointerUsingRightValue) {
```
Comment (ziobron) [inline module1/homework/make_unique/tests/make_unique.ut.cpp]:
```suggestion
TEST(MakeUniqueTest, ShouldCreatePointerUsingMoveConstructor) {
```

**Review by ziobron — CHANGES_REQUESTED** [review]
We should pass in both cases (l and r value) object of the same type. If sideTestValue_ are different, then different constructors (copy/move) were called, so we have a perfect forwarding.

File: module1/homework/make_unique/tests/make_unique.ut.cpp [hunk 7]
```diff
@@ -73,13 +73,16 @@ TEST(MakeUniqueTest, ShouldCreateUniquePointerForArrayType) {
...
     ASSERT_EQ(sideTestClassPtr->sideTestValue_, leftValue);
+    ASSERT_EQ(sideTestClassPtr->sideTestValue_, classToCopy.sideTestValue_);
 }
 
-TEST(MakeUniqueTest, ShouldCreatePointerUsingRightValue) {
+TEST(MakeUniqueTest, ShouldCreatePointerUsingMoveConstructor) {
     LRValueTestClass classToMove;
     auto sideTestClassPtr = cs::make_unique<LRValueTestClass>(std::move(classToMove));
     ASSERT_EQ(sideTestClassPtr->sideTestValue_, rightValue);
+    ASSERT_TRUE(sideTestClassPtr->sideTestValue_ != classToMove.sideTestValue_);
```
Comment (ziobron) [inline module1/homework/make_unique/tests/make_unique.ut.cpp]: classToMove object is in a mover-from state, so you cannot refer to it. It is undefined behavior. This last check here may break the test. Please remove it.

**Review by ziobron — CHANGES_REQUESTED** [review]

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
{ "requirement_ids": [1119, 4443], "points": [3, 40] }

**Pull request closed without merging**